keps/sig-cli/2227-kubectl-default-container/README.md

```diff
@@ -0,0 +1,300 @@
+
+If `kubectl.kubernetes.io/default-logs-container` is specified, we should use this annotation instead of the general one for `kubectl log` and use general annotation for other commands.
+
+### Notes/Constraints/Caveats
+
+As the annotation `kubectl.kubernetes.io/default-container` will not be automatically added, users and Pod owners will need to be aware of this new annotation.
+
+### Risks and Mitigations
+**Note:** No server-side changes are required for this, all Request and Response template expansion is performed on
+the client side.
+
+- None
+
+## Design Details
+
+**Publishing Data:**
+
+Alpha:  default container annotation
+
+- Define a well known annotation `kubectl.kubernetes.io/default-container` in a Pod to provide a way to consumers (CLI, operators) to known which is the default Container.
+- Define a global function GetDefaultContainerName for kubectl that uses this annotation like below
+```
+func GetDefaultContainerName(p *ExecOptions, pod *corev1.Pod) string{
+  containerName := p.ContainerName
+  if len(containerName) != 0 {
+    return containerName
+  }
+
+  if len(pod.Spec.Containers) > 1 {
+    // in case the "kubectl.kubernetes.io/default-container" annotation is present, we preset the opts.Containers to default to selected
+    // container. This gives users ability to preselect the most interesting container in pod.
+    if annotations := pod.Annotations; annotations != nil && len(annotations[corev1.DefaultContainerAnnotationName]) > 0 {
+      containerName := annotations[corev1.DefaultContainerAnnotationName]
+      if exists, _ := podutils.FindContainerByName(pod, containerName); exists != nil {
+        return containerName
+      } else {
+        fmt.Fprintf(p.ErrOut, "Default container name %q in annotation not found in a pod\n", containerName)
+      }
+    }
+    fmt.Fprintf(p.ErrOut, "Defaulting container name to first container %s.\n", pod.Spec.Containers[0].Name)
+
+    if p.EnableSuggestedCmdUsage {
+      fmt.Fprintf(p.ErrOut, "Use '%s describe pod/%s -n %s' to see all of the containers in this pod.\n", p.ParentCommandName, pod.Name, p.Namespace)
+    }
+  }
+
+  containerName = pod.Spec.Containers[0].Name
+  return containerName
+}
+```
+- Add test cases to make sure that the command is running with the right container. When `--container` is specified, the annotation will be ignored.
+- Validate the annotation value before using it, as container name should follow RFC 1123. If the annotation value is invalid or not found in the pod, a warning message is needed before exiting.
+- Add an Env 'KUBE_DEFAULT_CONTAINER_ANNOTATION=disabled', we will ignore the specified annotation. This Env is only for disabling this. By default, this feature should be enabled as this feature is opt-in, and it only works once user adds the specified annotation to their pods.
```

Comment:
We don't do that, specifying `--container` will override default behavior and that's sufficient. I prefer we don't introduce too many environment variables.